picasso-sample/src/main/java/com/example/picasso/SampleContactsActivity.java

```diff
@@ -40,9 +48,34 @@
     lv.setAdapter(adapter);
     lv.setOnScrollListener(new SampleScrollListener(this));
 
+    if (ActivityCompat.checkSelfPermission(this, READ_CONTACTS) == PERMISSION_GRANTED) {
+      loadContacts();
+    } else {
+      ActivityCompat.requestPermissions(this, new String[] { READ_CONTACTS },
+          REQUEST_READ_CONTACTS);
+    }
+  }
+
+  private void loadContacts() {
     getSupportLoaderManager().initLoader(ContactsQuery.QUERY_ID, null, this);
   }
 
+  @Override public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+      @NonNull int[] grantResults) {
+    if (requestCode == REQUEST_READ_CONTACTS) {
+      if (grantResults.length > 0
+          && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+        loadContacts();
+      } else {
+        Toast.makeText(getApplicationContext(), "Read contacts permission denied",
```

Comment:
to nitpick: just use `this` instead of the application.

Reply:
wouldn't that cause a leak, since the toast could outlive the activity, say, by hitting the back button?

Reply:
ah, i guess so. thought Toast already called getApplicationContext, but I'm wrong!

Reply:
Toasts immediately leave the process. Using `this` is fine.

Reply:
Updated